src/newgrf_badge.cpp

```diff
@@ -0,0 +1,564 @@
+		});
+	}
+
+	std::span<const BadgeClassID> Classes() const { return this->classes; }
+
+private:
+	std::vector<BadgeClassID> classes; ///< List of badge classes.
+};
+
+static bool operator<(const GUIBadgeClasses::Element &a, const GUIBadgeClasses::Element &b)
+{
+	if (a.column_group != b.column_group) return a.column_group < b.column_group;
+	if (a.sort_order != b.sort_order) return a.sort_order < b.sort_order;
+	return a.label < b.label;
+}
+
+/**
+ * Construct of list of badge classes and column groups to display.
+ * @param feature feature being used.
+ */
+GUIBadgeClasses::GUIBadgeClasses(GrfSpecFeature feature)
+{
+	/* Get list of classes used by feature. */
+	UsedBadgeClasses used(feature);
+
+	uint max_column = 0;
+	for (BadgeClassID class_index : used.Classes()) {
+		Dimension size = GetBadgeMaximalDimension(class_index, feature);
+		if (size.width == 0) continue;
+
+		uint8_t column = 0;
+		bool visible = true;
+		uint sort_order = UINT_MAX;
+
+		std::string_view label = GetClassBadge(class_index)->label;
+		ApplyBadgeClassConfiguration(feature, label, visible, column, sort_order);
+
+		this->gui_classes.emplace_back(class_index, column, visible, sort_order, size, label);
+		if (visible) max_column = std::max<uint>(max_column, column);
+	}
+
+	std::sort(std::begin(this->gui_classes), std::end(this->gui_classes));
+
+	/* Determine total width of visible badge columns. */
+	this->column_widths.resize(max_column + 1);
+	for (const auto &el : this->gui_classes) {
+		if (!el.visible) continue;
+		this->column_widths[el.column_group] += ScaleGUITrad(el.size.width) + WidgetDimensions::scaled.hsep_normal;
+	}
+
+	/* Replace trailing `hsep_normal` spacer with wider `hsep_wide` spacer. */
+	for (uint &badge_width : this->column_widths) {
+		if (badge_width == 0) continue;
+		badge_width = badge_width - WidgetDimensions::scaled.hsep_normal + WidgetDimensions::scaled.hsep_wide;
+	}
+}
+
+/**
+ * Get total width of all columns.
+ * @returns sum of all column widths.
+ */
+uint GUIBadgeClasses::GetTotalColumnsWidth() const
+{
+	return std::accumulate(std::begin(this->column_widths), std::end(this->column_widths), 0U);
+}
+
+/**
+ * Construct a badge text filter.
+ * @param filter string filter.
+ * @param feature feature being used.
+ */
+BadgeTextFilter::BadgeTextFilter(StringFilter &filter, GrfSpecFeature feature)
+{
+	/* Do not filter if the filter text box is empty */
+	if (filter.IsEmpty()) return;
+
+	/* Pre-build list of badges that match by string. */
+	for (const auto &badge : _badges.specs) {
+		if (badge.name == STR_NULL) continue;
+		if (!badge.HasFeature(feature)) continue;
+
+		filter.ResetState();
+		filter.AddLine(GetString(badge.name));
+		if (!filter.GetState()) continue;
+
+		auto it = std::ranges::lower_bound(this->badges, badge.index);
+		if (it != std::end(this->badges) && *it == badge.index) continue;
+
+		this->badges.insert(it, badge.index);
+	}
+}
+
+/**
+ * Test if one of the lists badges matches the filtered badge list.
```

Comment:
```suggestion
 * Test if one of the list's badges matches the filtered badge list.
```
Or maybe just 'of the given badges'?